### PR TITLE
Enable map configurations (config v2).

### LINKED ISSUE #414: Configuration v2
A bit of an RFC
## Issues

There's some "problems" with the current configuration and I'd like to address those in a backwards incompatible way (probably through a migration period where both formats are allowed).

There's two primary issues I'd like to address with the current configuration format:
- The top level is currently a list, making it difficult to add repository-level configuration: #240 #281 (as well as other things I'd like such as repository-level interpreter version)
- The configuration is not scriptably rewriteable: #210

There's also auxiliary issues:
- yaml is not maintained
- yaml has some security issues #396
- `hooks.yaml` isn't well namespaced
### A new configuration language

I shopped around for configuration languages which fit the criteria above, notably:
- Provides a non-C implementation
- Supports comments + maps + lists
- Supports round-trip rewritability
- Has a python implementaiton

And did not find anything.  Given that, I've started speccing and implementing [a configuration language](https://github.com/asottile/dumbconf) with those design goals in mind.  It also aims to be significantly simpler than yaml (the spec may not fit on a postcard like json, but it's also not 70+ pages like yaml).
### New filenames

The names of the files will need to change for two reasons:
- The configuration will no longer be yaml
- hooks.yaml is not a namespaced-enough name

Proposing:
- `.pre-commit-config` - describes the consumer configuration (previously `.pre-commit-config.yaml`)
- `.pre-commit-manifest` - describes the repository configuration (previously `hooks.yaml`)
### New configuration structure

The new configuration structure will be a top-level mapping type to allow repository-wide configuration.  This map will have a `hooks` key which will contain the current configuration.

``` yaml
hooks: [
        ...,
]
```

This will allow for top-level configuration which will apply to all hooks (such as global excludes, interpreter versions, etc.)
## Migration
- The new configuration will be introduced.
- In that release and the next two releases (or the next 3 months, whichever comes sooner), pre-commit will load both configuration formats but issue warnings for the legacy format.
- After the migration period, the old configuration format will be purged.

The "officially" maintained repositories ([pre-commit/pre-commit-hooks](https://github.com/pre-commit/pre-commit-hooks) as well as the [mirrors-*](https://github.com/pre-commit) repositories) will force-push tags for the lastest $n versions and provide both configuration formats for those tags.  Going forward, [pre-commit/pre-commit-mirror-maker](pre-commit/pre-commit-mirror-maker) will only produce the new configuration format.

To assist in migration (though the formats should be relatively easy to manually migrate) a tool will be created `pre-commit-migrate` which will attempt to take the legacy format and produce new configuration files.  Given yaml's lack of rewritability this will likely reformat lists and drop comments, but for the last time!

Thoughts?

### PR DESCRIPTION
**Pull request by asottile**
This resolves #414

I've also added the following to aid in migration:
- `pre-commit migrate-config`
    - attempts to intelligently upgrade the configuration to the new map format
- `pre-commit autoupdate` - by default calls `migrate_config(...)`

For now I have opted to _not_ issue a deprecation warning, I'll follow up with that in a later release.

### CODE REVIEW

**Comment by asottile** [comment]
This is mostly ready for review, I need to make `migrate-config` handle yaml that looks like this:

```yaml
---
-   repo: ...
    ...
```

And should probably handle this case better:

*current*
```diff
+repos:
 # this is a pre-commit config
 -   repo: ...
     ...
```
*better*
```diff
 # this is a pre-commit config
+repos:
 -   repo: ...
     ...
```

**Comment by asottile** [comment]
Now handles both of those cases!

File: pre_commit/clientlib.py [hunk 1]
```diff
@@ -125,7 +126,11 @@ def validate_manifest_main(argv=None):
         ensure_absent=True,
     ),
 )
-CONFIG_SCHEMA = schema.Array(CONFIG_REPO_DICT)
+CONFIG_SCHEMA = schema.Map(
+    'Config', None,
+
+    schema.RequiredRecurse('repos', schema.Array(CONFIG_REPO_DICT)),
```
Comment (chriskuehl) [inline pre_commit/clientlib.py]: weird blank line?

Comment (asottile) [reply pre_commit/clientlib.py]: this matches the other ones in the file, I plan to add more keys here so it'll look more natural _eventually_

**Pull request merged**